### PR TITLE
PDF417: Automated structured append

### LINKED ISSUE #92: PDF417: Automated structured append
It might be nice if there were a way to have Okapi automatically slice-and-dice data into multiple structured append (Macro PDF417) symbols.

### PR DESCRIPTION
**Pull request by uwolfer**
Resolves #92

### CODE REVIEW

**Comment by gredler** [comment]
Thanks, will review this week!

**Comment by uwolfer** [comment]
@gredler I have rebased this change and applied slight improvements. Please let me know what you think.

**Comment by gredler** [comment]
Will do -- this is high on my to-do list.

**Comment by uwolfer** [comment]
FYI, I did now also extensive manual testing. Generated a series of e-tax 27 barcodes with both J4L and Okapi with the same input, and compared the output. All generated barcodes are 100% equal (pixel-by-pixel).

**Comment by gredler** [comment]
I'm a little hesitant to coordinate this code review via PR comments, but let's give it a try and see how it goes :-)

One high-level item to start: Can you resolve the conflicts? It looks like I caused a one- or two-line conflict with my recent line ending fix.

**Comment by uwolfer** [comment]
> Can you resolve the conflicts?

Done.

**Review by gredler — CHANGES_REQUESTED** [review]
Reviewed original PR last week

File: src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                    return remainingContent.substring(mid);
+                }
+                high = mid - 1;
+            }
+        }
+
+        throw new RuntimeException("Failed to set content, this is a bug.");
+    }
+
+    private static SetContentStatus trySetContent(String remainingContent, Pdf417 symbol, int lengthToTry, boolean forceByteCompaction) {
```
[the remark this answers is not in the document]
Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: If I use the template instead of the derived symbol for testing max. content length, I could not override `plotSymbol` with an empty impl, which could lead in degraded performance if I remember correctly a past comment from your side. Or do I miss something here?

[the remark this answers is not in the document]
Comment (gredler) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: Yes, you want to use a "lite" Pdf417 subclass, but you don't necessarily need to build the "lite" clone based on the template until you're in the `trySetContent` / `fits` method. My idea was that this simplifies the outer loop so that the combined outer + inner loop (discussed in a different comment) would be easier to understand (once combined) since the cloning noise is pushed down into `fits`.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            clone(template, symbol);
+
+            symbol.setStructuredAppendPosition(contentList.size() + 1);
+            symbol.setStructuredAppendTotal(MAX_STRUCTURED_APPEND_TOTAL);
+
+            String currentContent = remainingContent;
+            remainingContent = setContentAndGetRemainingWithBinarySearch(currentContent, symbol, forceByteCompaction);
+            contentList.add(currentContent.replace(remainingContent, ""));
+        }
+        return contentList;
```
[the remark this answers is not in the document]
Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: Good catch! Also adjusted the test to cover this case.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: I was able to change this partially only: moved the symbol template handling into this method. But have not found a way yet to refactor this to a boolean returning method. I am quite sure there is such a way, but somehow I am struggling here right now. Do you mind creating a patch which shows your idea?

**Comment by uwolfer** [comment]
@gredler I have now added more test cases which cover boundary cases, and thus also found & fixes some issues uncovered by these.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: I gave it another try, but was not able to find a solution which was nearly as performing as the current implementation unfortunately.
@gredler If you still think we can simplify this method, I would appreciate if you could give it a try.

[the remark this answers is not in the document]
Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: @gredler Gave it another try, and was able to implement it. Please let me know if that is what you were thinking of.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                int endIndex = lastData.length() - 1;
+                dataList.set(dataList.size() - 1, lastData.substring(0, endIndex));
+                dataList.add(lastData.substring(endIndex));
+            }
+        }
+        return dataList;
+    }
+
+    private static boolean fits(String data, Pdf417 template, boolean last) {
+        Pdf417 symbol = new Pdf417() {
```
Comment (uwolfer) [inline src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java:96]: Instead of instantiating here, we could even instantiate out of the loops, and reuse the one instance for all the tests. Do you think that would be safe?

Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java]: I have changed it like that, when passing in huge amount of data, instantiating testing symbols showed up in profiling. Let me know in case you see issues with this approach.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            } // expensive plotting is not required
+        };
+        clone(template, testSymbol);
+        testSymbol.setStructuredAppendTotal(2);
+
+        List<String> dataList = new ArrayList<>();
+        String remainingData = data;
+        while (!remainingData.isEmpty()) {
+            int low = 0;
+            int high = remainingData.length();
```
Comment (uwolfer) [inline src/main/java/uk/org/okapibarcode/backend/Pdf417AutoStructuredAppend.java:73]: @gredler Do you think we can limit the `high`? I.e. is there a way to calculate the maximum ever possible data amount for a specified symbol? That would allow us to optimize performance.

**Comment by gredler** [comment]
I think the PR is looking good! I need to dig into the core of `splitData` a little more, but I think we're close. In the meantime, can you make a couple of small tweaks?

Let's move the implementation from `Pdf417AutoStructuredAppend` to `Pdf417` itself, at the bottom of the class. I know it's already a long class, but that's the Zint / C code legacy, and we have to tread a fine line between "clean" Java code and not deviating too much from the original code (to make issues easier to diagnose and resolve). The approach (consistent with the original C code) is to have one file / class per symbol, and maintaining that consistency is probably more valuable than making things "cleaner" in one small corner of the code.

The JavaDoc for the byte[] version mentions `setContent(byte[])`, which is not used. It probably doesn't need to get too specific about how it works under the covers.

I think we can rename `plotStructuredAppendSymbols` to `createStructuredAppendSymbols` for consistency with the other methods.

The copyright in the `Pdf417AutoStructuredAppendTest` file can be changed to your name.

I'll try to check that last little bit in `splitData` later today!

**Comment by uwolfer** [comment]
Thanks. Just pushed an update which addresses your comments.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417.java [hunk 5]
```diff
@@ -1847,4 +1848,129 @@ public String toString() {
...
+        testSymbol.setStructuredAppendTotal(2);
+
+        List<String> dataList = new ArrayList<>();
+        String remainingData = data;
+        while (!remainingData.isEmpty()) {
+            int low = 0;
+            int high = remainingData.length();
+
+            while (low <= high) {
+                int mid = low + high >>> 1;
```
Comment (gredler) [inline src/main/java/uk/org/okapibarcode/backend/Pdf417.java:1899]: What is the order of operations here? Is it shift *after* addition?

File: src/main/java/uk/org/okapibarcode/backend/Pdf417.java [hunk 6]
```diff
@@ -1847,4 +1848,129 @@ public String toString() {
...
+                int currentLength = Math.min(mid, remainingData.length());
```
Comment (gredler) [inline src/main/java/uk/org/okapibarcode/backend/Pdf417.java:1900]: Is this `Math.min()` necessary? I'm having trouble envisioning when `mid` would be larger than `remainingData.length()`.

Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417.java:1900]: Yes, in my understanding this is required. There might be a situation where `low` becomes greater than the actual length of the remaining data due to the increment in the binary search process. In such a case, `mid` would also be greater than the length of the remaining data.

File: src/main/java/uk/org/okapibarcode/backend/Pdf417.java (hunk 5 above)
Comment (uwolfer) [reply src/main/java/uk/org/okapibarcode/backend/Pdf417.java:1899]: Yes. Addition has higher precedence than the left shift and right shift operators ([source](https://www.javatpoint.com/highest-precedence-in-java#:~:text=The%20operator%20precedence%20is%20responsible,shift%20and%20Right%20shift%20operators.)).

File: src/main/java/uk/org/okapibarcode/backend/Pdf417.java [hunk 7]
```diff
@@ -1847,4 +1848,129 @@ public String toString() {
...
+            } // expensive plotting is not required
+        };
+        clone(template, testSymbol);
+        testSymbol.setStructuredAppendTotal(2);
+
+        List<String> dataList = new ArrayList<>();
+        String remainingData = data;
+        while (!remainingData.isEmpty()) {
+            int low = 0;
+            int high = remainingData.length();
```
Comment (uwolfer) [inline src/main/java/uk/org/okapibarcode/backend/Pdf417.java:1896]: @gredler Do you think we can limit the `high`? I.e. is there a way to calculate the maximum ever possible data amount for a specified symbol? That would allow us to optimize performance.

(Adding comment again there in order to make it visible after moving code around.)

**Review by gredler — APPROVED** [review]
